### PR TITLE
Improved pointer handling on UWP and Hover events

### PR DESCRIPTION
**Pull request by piersdeseilligny**
### Improved UWP Pointer handling

By correctly handling the `PointerEntered` and `PointerExited` events and capturing the pointer (`Container.CapturePointer(e.Pointer);`) in [TouchEffect.UWP/PlatformTouchEff.cs](https://github.com/piersdeseilligny/TouchEffect/blob/master/TouchEffect.UWP/PlatformTouchEff.cs), Touch effects created with this library will now behave in the exact same way as native UWP ones, which is especially noticeable when pressing the pointer down, and dragging the cursor outside the bounds of the control.

### Hover events for UWP and Android

As a "side-effect" of the above feature, I have started adding basic support for hover events. This is manifested as the new `TouchStatus.HoverEnter` and `TouchStatus.HoverLeave` statuses, which can be handled by users through the `StatusChanged` event. This feature can be previewed with the new HoverSample in the sample project.

Currently, `HoverEnter` and `HoverLeave` are only triggered on UWP and Android (although I have not tested it on Android, as I do not own a mouse to test it with). 
I have looked into adding support for Xamarin.Mac, however as far as I can tell this would require creating a subclass of NSView in order to override the `MouseEntered` and `MouseExited` events. Not being familiar with Cocoa, I don't really know how to approach this issue.

### CODE REVIEW

**Comment by AndreiMisiukevich** [comment]
Hi) This is a very cool enhancement! But I believe it should be achieved in a little bit different way.

1) separate enum HoverStatus { Started, Ended }
2) I don't think that Container_PointerExited or Container_PointerEnetered should raise TOUCH Canceled or Started events, but they will raise HoverStatus updates

Also, I think it would make sense to add HandleHover() method with own transformations stuff (usually it's a shadow/color changing)

Look how it works on mac on the web (Github). I like this behavior.

1) [Regular state] if no hover and no press
2) [Hovered state] if there is hover, but no press
3) [Pressed state] if there is hover and there is press (if the user ends press, we should call command)
3) [Pressed state] if no hover, BUT there is press ((if the user ends press, we should NOT call command. Canceled status)

So if the user pressed a button and shifted mouse outside of button boundaries we still show the button as pressed, but DON'T raise "Success" once he/she ends press

Here is the GIF.

Untitled

**Comment by piersdeseilligny** [comment]
For UWP, I implemented the behaviour which is seen in native UWP apps:

[Gif](https://gyazo.com/8b676de9dc5382c8c85bac335d60e4ec)

It's the same as what you described, except that moving the pointer outside of the control causes it to lose it's pressed state, however it regains this state if you move the cursor back in while the pointer is pressed. Does this not make it clearer that releasing the pointer outside the bounds of the control does NOT trigger the command?

I don't know what the native behavior is on other platforms though, maybe it would be best to imitate the native behavior of each platform?

**Comment by AndreiMisiukevich** [comment]
> For UWP, I implemented the behavior which is seen in native UWP apps:
> 
> [Gif](https://gyazo.com/8b676de9dc5382c8c85bac335d60e4ec)
> 
> It's the same as what you described, except that moving the pointer outside of the control causes it to lose it's pressed state, however it regains this state if you move the cursor back in while the pointer is pressed. Does this not make it clearer that releasing the pointer outside the bounds of the control does NOT trigger the command?
> 
> I don't know what the native behavior is on other platforms though, maybe it would be best to imitate the native behavior of each platform?


Totally agree)

Just checked on native mac app (now browser)


1) [Regular state] if no hover and no press
2) [Hovered state] if there is hover, but no press
3) [Pressed state] if there is hover and there is press (if the user ends press, we should call command)
4) [**Regular** state] if no hover, BUT there is press ((if the user ends press, we should NOT call command. Canceled status)

**Comment by AndreiMisiukevich** [comment]
but anyway we should keep hower and touch separately
Do you agree?

**Comment by piersdeseilligny** [comment]
> but anyway we should keep hower and touch separately
> Do you agree?

Yep that sounds good. If you want I can work on separating the hover and touch logic, however I can't work on adding the hover functionality to Mac, seeing as I'm not at all familiar with developing for it, and don't even own a macOS machine.

**Comment by AndreiMisiukevich** [comment]
@piersdeseilligny let's split work for several steps

1st step: 
- Create new enum HoverStatus { Exited, Entered }
- Handle **UWP** events OnPointerExited and OnPointerEntered with passing corresponding statuses. (no inRange check on platform side. all these calculations will be in crossplaftorm code i believe)

Then decide what's next :)

And please don't update XF and all other files. Only necessary things should be included

Feel free to open new PR or modify current

**Comment by AndreiMisiukevich** [comment]
btw, i am not sure if it makes sense to keep _pressed in UWP. It was added by a contributor. But is it needed indeed?

**Comment by piersdeseilligny** [comment]
Yeah _pressed is necessary to get the right behaviour when pressing the pointer, and exiting and re-entering the control. Although if all the inrange checks are moved to the crossplatform code, then it might not be necessary. However are we sure this is the best thing to do? Because all platforms handle the hover events slightly different (for example windows fires them even when using a touchscreen, while android doesn't) so I think it might be easier to get consistent behaviour by implementing hover individually for each platform? Although I'm not entirely sure yet; i'il look at it more closely later today.

**Comment by AndreiMisiukevich** [comment]
> Yeah _pressed is necessary to get the right behaviour when pressing the pointer, and exiting and re-entering the control. Although if all the inrange checks are moved to the crossplatform code, then it might not be necessary. However are we sure this is the best thing to do? Because all platforms handle the hover events slightly different (for example windows fires them even when using a touchscreen, while android doesn't) so I think it might be easier to get consistent behaviour by implementing hover individually for each platform? Although I'm not entirely sure yet; i'il look at it more closely later today.

I think we can start with crossplatform solution, then provide possibility to override it per platform if needed. For now, I like UWP approach

**Comment by piersdeseilligny** [comment]
So I've looked over it some more, and I think it makes sense to have the `_inrange` stuff on the platform-specific code. I think the code will be less messy, because some platforms (ios) don't even trigger hover, so there would be no way of knowing if `_inrange` is true, unless we add platform conditional code in the crossplatform part, at which point might as well make it platform-specific to start with.

Additionally, I think the codebase is more readable if the `HandleTouch` TouchStatus events triggered in the platform specific code are the same ones which make it to the TouchStatusChanged event on the other side.

If we adhere to the following guidelines, we should get consistent behavior on each platform:

* Pressing the pointer down and dragging it outside the bounds of the control triggers `TouchStatus.Canceled`

    * If it is then released outside the bounds of the control, nothing happens
    * If it reenters the bounds of the control, **before being released**, `TouchStatus.Started` is triggered again.

* If the platform supports it, `HoverStatus` is set to `Hovering` every time a pointer **enters** the bounds of the control. However it must always occur **before** any change to `TouchStatus` triggered by the same pointer action.

* If the platform supports it, `HoverStatus` is set to `Regular` every time a pointer **exits** the bounds of the control. However it must always occur **after** any change to `TouchStatus` triggered by the same pointer action.

**Comment by piersdeseilligny** [comment]
So I've added everything HoverState and HoverStatus related. Additionally, I have added an `_inrange` check to Android in order to get the same behavior as on UWP when dragging the pointer outside the bounds of the control while it is pressed.

What I'm guessing the next steps are is:

* **UWP**
    * Nothing, it's pretty much finished

* **Android**
    * Test hover with a mouse connected to an android device

* **iOS**
    * Nothing, hover is not supported. Maybe an `_inrange` check would be useful though? I don't have a mac, so can't really develop anything for iOS though.

* **Mac**
    * Figure out how to use `NSTrackingArea` without creating a subclass of `NSView` (if possible?). Then, implement all hover-related events.
    * Like for iOS, I can't really work on this, because I don't actually own a mac :/ 

* **Cross-Platform**
    * Add HoverBackgroundColor, HoverScale, HoverOpacity, HoverRotation, HoverRotationX, HoverRotationY, HoverTranslationX, HoverTranslationY, HoverAnimationDuration to `TouchEff.cs`
    * Add HoverBackgroundImageSource and HoverBackgroundImageAspect to `TouchImage.cs`

**Comment by AndreiMisiukevich** [comment]
@piersdeseilligny can you please clean the PR?
I mean remove all files which are not related to this PR

(Resource files, packages updates etc.)

File: TouchEffectSample/TouchEffectSample.csproj [hunk 1]
```diff
@@ -5,7 +5,7 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="Xamarin.Forms" Version="4.2.0.848062" />
+    <PackageReference Include="Xamarin.Forms" Version="4.3.0.947036" />
```
Comment (AndreiMisiukevich) [inline TouchEffectSample/TouchEffectSample.csproj]: should be excluded

File: TouchEffectSample.sln [hunk 2]
```diff
@@ -1,6 +1,6 @@
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 15
-VisualStudioVersion = 15.0.28307.136
+# Visual Studio Version 16
+VisualStudioVersion = 16.0.29418.71
```
Comment (AndreiMisiukevich) [inline TouchEffectSample.sln]: should be excluded

File: TouchEffectSample.sln [hunk 3]
```diff
@@ -96,6 +96,7 @@ Global
 		{9EED875F-F3C4-4753-9064-39DF7CD222AA}.AppStore|x86.Deploy.0 = Release|Any CPU
 		{9EED875F-F3C4-4753-9064-39DF7CD222AA}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
 		{9EED875F-F3C4-4753-9064-39DF7CD222AA}.Debug|Any CPU.Build.0 = Debug|Any CPU
+		{9EED875F-F3C4-4753-9064-39DF7CD222AA}.Debug|Any CPU.Deploy.0 = Debug|Any CPU
```
Comment (AndreiMisiukevich) [inline TouchEffectSample.sln]: should be excluded

File: TouchEffectSample.iOS/TouchEffectSample.iOS.csproj [hunk 4]
```diff
@@ -148,7 +148,7 @@
   </ItemGroup>
```
Comment (AndreiMisiukevich) [inline TouchEffectSample.iOS/TouchEffectSample.iOS.csproj]: the file should be discarded

File: TouchEffectSample.UWP/TouchEffectSample.UWP.csproj [hunk 5]
```diff
@@ -159,7 +159,7 @@
       <Version>6.2.2</Version>
     </PackageReference>
     <PackageReference Include="Xamarin.Forms">
-      <Version>3.6.0.344457</Version>
+      <Version>4.3.0.947036</Version>
```
Comment (AndreiMisiukevich) [inline TouchEffectSample.UWP/TouchEffectSample.UWP.csproj]: update pls to current project version (not latest on NuGet)

File: TouchEffectSample.Android/TouchEffectSample.Android.csproj [hunk 6]
```diff
@@ -119,5 +118,5 @@
       <Name>TouchEffect</Name>
     </ProjectReference>
   </ItemGroup>
- <Import Project="$(MSBuildExtensionsPath)\Xamarin\Android\Xamarin.Android.CSharp.targets" />
-</Project>
```
Comment (AndreiMisiukevich) [inline TouchEffectSample.Android/TouchEffectSample.Android.csproj]: the file should be discarded

File: TouchEffectSample.Android/Resources/Resource.designer.cs [hunk 7]
```diff
@@ -107,8 +107,8 @@ public static void UpdateIdValues()
 			global::TouchEffect.Droid.Resource.Attribute.behavior_overlapTop = global::TouchEffectSample.Droid.Resource.Attribute.behavior_overlapTop;
 			global::TouchEffect.Droid.Resource.Attribute.behavior_peekHeight = global::TouchEffectSample.Droid.Resource.Attribute.behavior_peekHeight;
 			global::TouchEffect.Droid.Resource.Attribute.behavior_skipCollapsed = global::TouchEffectSample.Droid.Resource.Attribute.behavior_skipCollapsed;
-			global::TouchEffect.Droid.Resource.Attribute.borderWidth = global::TouchEffectSample.Droid.Resource.Attribute.borderWidth;
 			global::TouchEffect.Droid.Resource.Attribute.borderlessButtonStyle = global::TouchEffectSample.Droid.Resource.Attribute.borderlessButtonStyle;
+			global::TouchEffect.Droid.Resource.Attribute.borderWidth = global::TouchEffectSample.Droid.Resource.Attribute.borderWidth;
```
Comment (AndreiMisiukevich) [inline TouchEffectSample.Android/Resources/Resource.designer.cs:111]: the file should be discarded

File: TouchEffect/TouchVisualManager.cs [hunk 8]
```diff
@@ -97,6 +97,24 @@ public void HandleTouch(ITouchEff sender, TouchStatus status)
             }
         }
 
+        public void HandleHover(ITouchEff sender, HoverStatus status)
+        {
+            if (sender.Control.IsEnabled)
```
Comment (AndreiMisiukevich) [inline TouchEffect/TouchVisualManager.cs]: please use the "guard" pattern and null-check

```csharp
if (!(sender.Control?.IsEnabled ?? false)) {
    return;
}

File: TouchEffect/TouchVisualManager.cs [hunk 9]
```diff
@@ -97,6 +97,24 @@ public void HandleTouch(ITouchEff sender, TouchStatus status)
...
+            {
+                sender.HoverStatus = status;
+                sender.RaiseHoverStatusChanged();
+                if (status == HoverStatus.Entered && sender.HoverState== HoverState.Regular)
```
Comment (AndreiMisiukevich) [inline TouchEffect/TouchVisualManager.cs]: space after sender.HoverState

File: TouchEffect/TouchVisualManager.cs [hunk 10]
```diff
@@ -97,6 +97,24 @@ public void HandleTouch(ITouchEff sender, TouchStatus status)
...
+                {
+                    sender.HoverState = HoverState.Hovering;
+                    sender.RaiseHoverStateChanged();
+                }
+                else if(status== HoverStatus.Exited && sender.HoverState == HoverState.Hovering)
```
Comment (AndreiMisiukevich) [inline TouchEffect/TouchVisualManager.cs]: space after status

**Review by AndreiMisiukevich — CHANGES_REQUESTED** [review]
Do we really need both (HoverStatus and HoverState)? They both have 2 "states"

**Comment by AndreiMisiukevich** [comment]
@piersdeseilligny overall: good job! 👍

File: TouchEffectSample.iOS/TouchEffectSample.iOS.csproj (hunk 4 above)
Comment (piersdeseilligny) [reply TouchEffectSample.iOS/TouchEffectSample.iOS.csproj]: I think I'm missing something, but why? Wouldn't this break the iOS Sample project?

File: TouchEffectSample.Android/TouchEffectSample.Android.csproj (hunk 6 above)
Comment (piersdeseilligny) [reply TouchEffectSample.Android/TouchEffectSample.Android.csproj]: Once again, think I'm missing something here? Do you mean the `TouchEffectSample.Android.csproj` file?

File: TouchEffectSample.Android/Resources/Resource.designer.cs (hunk 7 above)
Comment (piersdeseilligny) [reply TouchEffectSample.Android/Resources/Resource.designer.cs]: Isn't resource.designer.cs automatically generated?

**Comment by piersdeseilligny** [comment]
I've reverted my initial `Update Xamarin.Forms` commit, but now I have all the build errors which caused me to update everything in the first place. Although I'm not sure if it's just issues with my specific environment or not.

Would you mind handling this on your side of things?

**Comment by piersdeseilligny** [comment]
> Do we really need both (HoverStatus and HoverState)? They both have 2 "states"

I added both of them in for consistency with the `TouchStatus` and `TouchState`, one of them being a state, while the other is more of an event. 

In a multi-touch situation, HoverStatus is updated any time a pointer enters the bounds of the control, while HoverState is only updated if the overall "hover state" changes. Also, `HoverStatus` might be useful in the future, if we want to add additional hover-related events?

File: TouchEffect/TouchEffect.csproj [hunk 11]
```diff
@@ -8,9 +8,6 @@
     <PackageReference Include="Xamarin.Forms" Version="4.2.0.848062" />
   </ItemGroup>
   <ItemGroup>
-    <Folder Include="Delegates\" />
-    <Folder Include="EventArgs\" />
-    <Folder Include="Enums\" />
```
Comment (AndreiMisiukevich) [inline TouchEffect/TouchEffect.csproj]: why removing them?

Comment (piersdeseilligny) [reply TouchEffect/TouchEffect.csproj]: Just a mistake, thanks for pointing it out

**Comment by piersdeseilligny** [comment]
So I'm pretty sure everything is fine now: Only the relevant files are included in this pull request, it builds fine, and seems free of bugs.

File: TouchEffect.Droid/PlatformTouchEff.cs [hunk 12]
```diff
@@ -58,19 +58,75 @@ protected override void OnDetached()
...
         private void OnTouch(object sender, AView.TouchEventArgs e)
         {
+            AView senderView = sender as AView;
+
             switch (e.Event.ActionMasked)
             {
                 case MotionEventActions.Down:
-                    Element.GetTouchEff().HandleTouch(TouchStatus.Started);
-                    break;
+                    {
```
Comment (AndreiMisiukevich) [inline TouchEffect.Droid/PlatformTouchEff.cs]: no need of parentheses (remove all please in the file)

File: TouchEffect.Droid/PlatformTouchEff.cs [hunk 13]
```diff
@@ -58,19 +58,75 @@ protected override void OnDetached()
             }
         }
 
+        bool _inrange;
+        bool _pressed;
+        bool _leftrange = false;
```
Comment (AndreiMisiukevich) [inline TouchEffect.Droid/PlatformTouchEff.cs]: no need to set "false" as it's default value for bool.
And what is the purpose of this flag? we have _inRange (this is the same, no?)

File: TouchEffect.Droid/PlatformTouchEff.cs [hunk 14]
```diff
@@ -58,19 +58,75 @@ protected override void OnDetached()
             }
         }
 
+        bool _inrange;
```
Comment (AndreiMisiukevich) [inline TouchEffect.Droid/PlatformTouchEff.cs]: inRange please

File: TouchEffect.Droid/PlatformTouchEff.cs [hunk 15]
```diff
@@ -58,19 +58,75 @@ protected override void OnDetached()
...
+        bool _pressed;
+        bool _leftrange = false;
+        int[] twoIntArray = new int[2];
```
Comment (AndreiMisiukevich) [inline TouchEffect.Droid/PlatformTouchEff.cs]: please use _ also use "private" keyword
and choose a more suitable name for twoIntArray

File: TouchEffect.Droid/PlatformTouchEff.cs [hunk 16]
```diff
@@ -58,19 +58,75 @@ protected override void OnDetached()
...
         private void OnTouch(object sender, AView.TouchEventArgs e)
         {
+            AView senderView = sender as AView;
```
Comment (AndreiMisiukevich) [inline TouchEffect.Droid/PlatformTouchEff.cs]: var senderView = sender as AView;

File: TouchEffect.Mac/PlatformTouchEff.cs [hunk 17]
```diff
@@ -41,15 +41,21 @@ protected override void OnDetached()
         }
     }
 
+
```
Comment (AndreiMisiukevich) [inline TouchEffect.Mac/PlatformTouchEff.cs]: remove this empty space please

File: TouchEffect.UWP/PlatformTouchEff.cs [hunk 18]
```diff
@@ -6,6 +6,7 @@
 using Xamarin.Forms;
 using Xamarin.Forms.Internals;
 using Xamarin.Forms.Platform.UWP;
+using System.Diagnostics;
```
Comment (AndreiMisiukevich) [inline TouchEffect.UWP/PlatformTouchEff.cs]: please remove

File: TouchEffect.UWP/PlatformTouchEff.cs [hunk 19]
```diff
@@ -34,9 +36,27 @@ protected override void OnAttached()
                 Container.PointerPressed += OnPointerPressed;
                 Container.PointerReleased += OnPointerReleased;
                 Container.PointerCanceled += OnPointerCanceled;
+                Container.PointerExited += Container_PointerExited;
```
Comment (AndreiMisiukevich) [inline TouchEffect.UWP/PlatformTouchEff.cs]: use consistent naming OnPointerExited

File: TouchEffect.UWP/PlatformTouchEff.cs [hunk 20]
```diff
@@ -34,9 +36,27 @@ protected override void OnAttached()
...
+                Container.PointerEntered += Container_PointerEntered;
```
Comment (AndreiMisiukevich) [inline TouchEffect.UWP/PlatformTouchEff.cs]: OnPointerEntered

File: TouchEffect/Delegates/TEffectHoverStateChangedHandler.cs [hunk 21]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+using TouchEffect.EventArgs;
+using Xamarin.Forms;
+
+namespace TouchEffect.Delegates
+{
+    [Obsolete]
+    public delegate void TouchViewHoverStateChangedHandler(TouchView sender, HoverStateChangedEventArgs args);
```
Comment (AndreiMisiukevich) [inline TouchEffect/Delegates/TEffectHoverStateChangedHandler.cs]: That's a little weird to add obsolete API :)
please remove obsolete code at all (don't add it)

File: TouchEffect.Droid/PlatformTouchEff.cs (hunk 13 above)
Comment (piersdeseilligny) [reply TouchEffect.Droid/PlatformTouchEff.cs]: It's indeed pointless, apologies (as was `_pressed` in fact)

File: TouchEffect/Delegates/TEffectHoverStatusChangedHandler.cs [hunk 22]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+using TouchEffect.EventArgs;
+using Xamarin.Forms;
+
+namespace TouchEffect.Delegates
+{
+    [Obsolete]
+    public delegate void TouchViewHoverStatusChangedHandler(TouchView sender, HoverStatusChangedEventArgs args);
```
Comment (AndreiMisiukevich) [inline TouchEffect/Delegates/TEffectHoverStatusChangedHandler.cs]: same here. just remove it

File: TouchEffect/Enums/TouchStatus.cs [hunk 23]
```diff
@@ -5,6 +5,6 @@ public enum TouchStatus
     {
         Started,
         Completed,
-        Canceled
+        Canceled,
```
Comment (AndreiMisiukevich) [inline TouchEffect/Enums/TouchStatus.cs]: remove comma (no need to change this file)

File: TouchEffect/TouchEff.cs [hunk 24]
```diff
@@ -511,6 +558,14 @@ public void RaiseStateChanged()
         public void RaiseStatusChanged()
             => StatusChanged?.Invoke(Control, new TouchStatusChangedEventArgs(Status));
 
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public void RaiseHoverStateChanged()
+    => HoverStateChanged?.Invoke(Control, new HoverStateChangedEventArgs(HoverState));
```
Comment (AndreiMisiukevich) [inline TouchEffect/TouchEff.cs]: fix tab space please

File: TouchEffectSample.Android/Resources/Resource.designer.cs (hunk 7 above)
Comment (AndreiMisiukevich) [inline TouchEffectSample.Android/Resources/Resource.designer.cs:111]: please discard this file as well

File: TouchEffectSample.sln (hunk 2 above)
Comment (AndreiMisiukevich) [inline TouchEffectSample.sln]: can you discard this file as well?

File: TouchEffectSample/HoverPage.xaml.cs [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+    public partial class HoverPage : ContentPage
+    {
+        public HoverPage()
+        {
+            InitializeComponent();
+        }
+
+        private void TouchEff_HoverStateChanged(VisualElement sender, TouchEffect.EventArgs.HoverStateChangedEventArgs args)
+        {
+            if (args.State == TouchEffect.Enums.HoverState.Hovering)
```
Comment (AndreiMisiukevich) [inline TouchEffectSample/HoverPage.xaml.cs]: sender.Opacity = args.State == TouchEffect.Enums.HoverState.Hovering ? 0.5 : 1.0;

**Review by AndreiMisiukevich — CHANGES_REQUESTED** [review]
final updates :)

**Review by AndreiMisiukevich — APPROVED** [review]

**Comment by AndreiMisiukevich** [comment]
well done
I will take a look at the weekend

**Comment by MitchBomcanhao** [comment]
Hover states on UWP will be very useful, thanks for your efforts!

**Pull request merged**